File: app/services/statement_processor.py
"""
Async statement processing pipeline:
upload → extract → classify → insert → detect_recurring → validate_balance
"""
from __future__ import annotations

import hashlib
import logging
from pathlib import Path

from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, update

from app.models.statement import Statement
from app.models.transaction import Transaction
from app.services.ml.classifier import classifier
from app.services.ml.recurring_detector import recurring_detector
from app.services.extraction.pdf_extractor import pdf_extractor
from app.services.extraction.csv_extractor import csv_extractor
from app.services.extraction.excel_extractor import excel_extractor
import uuid

logger = logging.getLogger(__name__)

MIME_EXTRACTOR_MAP = {
    "application/pdf": "pdf",
    "text/csv": "csv",
    "application/vnd.ms-excel": "excel",
    "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet": "excel",
    "text/plain": "csv",
}


def _make_hash(date: str, amount: str, description: str, tx_type: str) -> str:
    raw = f"{date}|{amount}|{description[:50]}|{tx_type}"
    return hashlib.sha256(raw.encode()).hexdigest()[:32]


async def _persist_subscriptions(db: AsyncSession, user_id: str, detected: list) -> None:
    """Upsert auto-detected subscriptions by name (avoids duplicates on re-upload)."""
    from app.models.subscription import Subscription

    existing = await db.execute(
        select(Subscription).where(Subscription.user_id == uuid.UUID(user_id))
    )
    by_name = {s.name.lower(): s for s in existing.scalars().all()}

    for d in detected:
        key = d.name.lower()
        if key in by_name:
            sub = by_name[key]
            sub.amount = d.amount
            sub.frequency = d.frequency
            sub.last_seen = d.last_seen
            sub.next_expected = d.next_expected
            sub.annual_cost = d.annual_cost
            sub.confidence = d.confidence
            sub.is_active = True
        else:
            db.add(Subscription(
                id=uuid.uuid4(),
                user_id=uuid.UUID(user_id),
                name=d.name,
                merchant=d.merchant,
                amount=d.amount,
                frequency=d.frequency,
                subcategory=d.subcategory,
                last_seen=d.last_seen,
                next_expected=d.next_expected,
                annual_cost=d.annual_cost,
                is_auto_detected=True,
                confidence=d.confidence,
            ))
    await db.commit()


async def process_statement(
    statement_id: str,
    file_path: str,
    mime_type: str,
    user_id: str,
    account_id: str | None,
) -> None:
    """Runs as a background task with its own DB session (the request's
    session is already closed by the time this executes)."""
    from app.database import AsyncSessionLocal
    async with AsyncSessionLocal() as db:
        await _process_statement_inner(
            statement_id, file_path, mime_type, user_id, account_id, db
        )


async def _process_statement_inner(
    statement_id: str,
    file_path: str,
    mime_type: str,
    user_id: str,
    account_id: str | None,
    db: AsyncSession,
) -> None:
    try:
        await db.execute(
            update(Statement)
            .where(Statement.id == uuid.UUID(statement_id))
            .values(status="processing")
        )
        await db.commit()

        # 1. Extract
        extractor_type = MIME_EXTRACTOR_MAP.get(mime_type, "csv")
        if extractor_type == "pdf":
            result = pdf_extractor.extract(file_path)
        elif extractor_type == "excel":
            result = excel_extractor.extract(file_path)
        else:
            result = csv_extractor.extract(file_path)

        if not result.transactions:
            await db.execute(
                update(Statement)
                .where(Statement.id == uuid.UUID(statement_id))
                .values(status="failed", error_message="No transactions could be extracted")
            )
            await db.commit()
            return

        # 2. Classify + build Transaction objects
        tx_objects: list[Transaction] = []
        seen_hashes: set[str] = set()

        # Check existing hashes for this user to prevent duplicates
        existing = await db.execute(
            select(Transaction.hash).where(Transaction.user_id == uuid.UUID(user_id))
        )
        db_hashes = {row[0] for row in existing.fetchall() if row[0]}

        for tx in result.transactions:
            clf = classifier.classify(tx.description, tx.amount)
            merchant = clf.merchant or classifier.extract_merchant(tx.description)

            tx_hash = _make_hash(str(tx.date), str(tx.amount), tx.description, tx.type)
            is_duplicate = tx_hash in db_hashes or tx_hash in seen_hashes
            seen_hashes.add(tx_hash)

            is_bank_charge = clf.subcategory in ("Bank Charges",)
            is_subscription = clf.category == "Subscriptions"

            obj = Transaction(
                id=uuid.uuid4(),
                user_id=uuid.UUID(user_id),
                account_id=uuid.UUID(account_id) if account_id else None,
                statement_id=uuid.UUID(statement_id),
                raw_description=tx.description,
                raw_date=tx.raw_date,
                raw_amount=tx.raw_amount,
                raw_balance=tx.raw_balance,
                description=tx.description,
                date=tx.date,
                amount=tx.amount,
                balance=tx.balance,
                currency="NGN",
                type=tx.type,
                category=clf.category if clf.category != "Unknown" else None,
                subcategory=clf.subcategory or None,
                merchant=merchant,
                confidence=clf.confidence,
                classified_by=clf.classified_by,
                is_recurring=False,
                is_subscription=is_subscription,
                is_bank_charge=is_bank_charge,
                is_duplicate=is_duplicate,
                hash=tx_hash,
            )
            tx_objects.append(obj)

        # 3. Batch insert
        BATCH = 100
        for i in range(0, len(tx_objects), BATCH):
            db.add_all(tx_objects[i:i + BATCH])
            await db.commit()

        # 4. Detect recurring subscriptions
        tx_dicts = [
            {
                "type": t.type,
                "amount": t.amount,
                "date": t.date,
                "description": t.description,
                "merchant": t.merchant,
                "category": t.category,
            }
            for t in tx_objects if not t.is_duplicate
        ]
        detected = recurring_detector.detect(tx_dicts)
        await _persist_subscriptions(db, user_id, detected)

        # 5. Update statement
        non_dup = [t for t in tx_objects if not t.is_duplicate]
        updates: dict = {
            "status": "completed",
            "transaction_count": len(non_dup),
        }
        if result.bank_name:
            updates["bank_name"] = result.bank_name
        if result.period_from:
            updates["period_from"] = result.period_from
        if result.period_to:
            updates["period_to"] = result.period_to
        if result.opening_balance:
            updates["opening_balance"] = result.opening_balance
        if result.closing_balance:
            updates["closing_balance"] = result.closing_balance

        await db.execute(
            update(Statement)
            .where(Statement.id == uuid.UUID(statement_id))
            .values(**updates)
        )
        await db.commit()

        logger.info(f"Statement {statement_id}: {len(non_dup)} transactions processed ({len(tx_objects) - len(non_dup)} duplicates skipped)")

    except Exception as e:
        logger.error(f"Statement processing failed for {statement_id}: {e}", exc_info=True)
        try:
            await db.rollback()
            await db.execute(
                update(Statement)
                .where(Statement.id == uuid.UUID(statement_id))
                .values(status="failed", error_message=str(e)[:500])
            )
            await db.commit()
        except Exception:
            pass

File: src/services/statementProcessor.ts
import fs from "fs/promises";
import path from "path";
import crypto from "crypto";
import { db } from "../db";
import {
  statements,
  transactions,
  subscriptions,
  subscriptionOccurrences,
  entities,
  entityAliases,
  unknownAccounts,
  categories,
} from "../db/schema";
import { eq, and } from "drizzle-orm";
import { extractFromCSV } from "./extraction/csvExtractor";
import { extractFromXLSX } from "./extraction/xlsxExtractor";
import { extractFromPDF } from "./extraction/pdfExtractor";
import { classifier } from "./ml/classifier";
import { recurringDetector } from "./ml/recurringDetector";
import { parseDate } from "./utils/parseHelpers";
import type { ProcessingJob, NormalizedTransaction, RawTransaction } from "../types";

export async function processStatement(job: ProcessingJob): Promise<void> {
  const { statementId, userId, filePath, fileType, accountId } = job;

  try {
    // Mark as processing
    await db
      .update(statements)
      .set({ status: "processing", processingStartedAt: new Date() })
      .where(eq(statements.id, statementId));

    // Read file
    const fileBuffer = await fs.readFile(filePath);
    const fileContent = fileBuffer.toString("utf-8");

    // Extract transactions
    let extractionResult;
    switch (fileType.toLowerCase()) {
      case "csv":
        extractionResult = await extractFromCSV(fileContent);
        break;
      case "xlsx":
      case "xls":
        extractionResult = await extractFromXLSX(fileBuffer);
        break;
      case "pdf":
      case "scanned_pdf":
        extractionResult = await extractFromPDF(fileBuffer);
        break;
      default:
        throw new Error(`Unsupported file type: ${fileType}`);
    }

    if (extractionResult.errors.length > 0 && extractionResult.transactions.length === 0) {
      await db
        .update(statements)
        .set({
          status: "failed",
          processingError: extractionResult.errors.join("; "),
        })
        .where(eq(statements.id, statementId));
      return;
    }

    // Update statement with period info
    await db
      .update(statements)
      .set({
        status: "extracted",
        periodStart: extractionResult.periodStart,
        periodEnd: extractionResult.periodEnd,
        openingBalance: extractionResult.openingBalance
          ? String(extractionResult.openingBalance)
          : undefined,
        closingBalance: extractionResult.closingBalance
          ? String(extractionResult.closingBalance)
          : undefined,
        transactionCount: extractionResult.transactions.length,
        extractionWarnings: extractionResult.warnings,
      })
      .where(eq(statements.id, statementId));

    // Classify and normalize transactions
    const rawTransactions = extractionResult.transactions;
    const normalizedTransactions: NormalizedTransaction[] = [];
    const transactionHashes = new Set<string>();

    // Fetch system categories for mapping
    const systemCats = await db
      .select()
      .from(categories)
      .where(eq(categories.isSystem, true));

    const categoryMap = new Map(
      systemCats.map((c) => [c.name.toLowerCase(), c.id])
    );

    const insertedTransactions = [];

    for (const raw of rawTransactions) {
      try {
        const date = parseDate(raw.date);
        if (!date) continue;

        const amount =
          raw.amount ||
          raw.debit ||
          raw.credit ||
          0;
        if (!amount || amount <= 0) continue;

        const type = raw.type || (raw.credit && raw.credit > 0 ? "credit" : "debit");
        const description = (raw.description || "").trim();

        if (!description) continue;

        // Deduplication hash
        const hashContent = `${date.toISOString().slice(0, 10)}|${amount.toFixed(2)}|${description.toLowerCase().slice(0, 50)}|${type}`;
        const txHash = crypto.createHash("sha256").update(hashContent).digest("hex").slice(0, 32);

        const isDuplicate = transactionHashes.has(txHash);
        transactionHashes.add(txHash);

        // Classify
        const classification = classifier.classify(description);

        const categoryId = classification.category
          ? categoryMap.get(classification.category.toLowerCase())
          : undefined;

        // Build transaction record
        const txData = {
          userId,
          statementId,
          accountId: accountId || undefined,
          rawDescription: description,
          rawDate: raw.date?.toString(),
          rawAmount: String(amount),
          rawBalance: raw.balance ? String(raw.balance) : undefined,
          date,
          amount: String(amount),
          currency: extractionResult.currency || "NGN",
          type: type as "debit" | "credit" | "transfer",
          description,
          balance: raw.balance ? String(raw.balance) : undefined,
          reference: raw.reference || undefined,
          merchant: classification.merchant || undefined,
          subcategory: classification.subcategory || undefined,
          confidence: classification.confidence
            ? String(classification.confidence)
            : "0",
          classifiedBy: classification.classifiedBy,
          categoryId: categoryId || undefined,
          isRecurring: classification.isRecurring || false,
          isSubscription: classification.isSubscription || false,
          isBankCharge: classification.isBankCharge || false,
          isDuplicate,
          isUncertain: classification.confidence < 0.5,
          transactionHash: txHash,
        };

        const normalized: NormalizedTransaction = {
          date,
          amount,
          currency: extractionResult.currency || "NGN",
          type: type as "debit" | "credit" | "transfer",
          description,
          rawDescription: description,
          merchant: classification.merchant,
          entity: classification.entity,
          category: classification.category,
          subcategory: classification.subcategory,
          confidence: classification.confidence,
          isRecurring: classification.isRecurring || false,
          isSubscription: classification.isSubscription || false,
          isBankCharge: classification.isBankCharge || false,
          isDuplicate,
        };

        normalizedTransactions.push(normalized);
        insertedTransactions.push(txData);
      } catch (err) {
        console.error("Failed to process transaction:", err);
      }
    }

    // Batch insert transactions
    if (insertedTransactions.length > 0) {
      const BATCH_SIZE = 100;
      for (let i = 0; i < insertedTransactions.length; i += BATCH_SIZE) {
        const batch = insertedTransactions.slice(i, i + BATCH_SIZE);
        await db.insert(transactions).values(batch);
      }
    }

    // Detect recurring subscriptions
    const debitTransactions = normalizedTransactions.filter(
      (t) => t.type === "debit"
    );
    const detectedSubs = recurringDetector.detect(debitTransactions);

    for (const sub of detectedSubs) {
      const existingSub = await db
        .select()
        .from(subscriptions)
        .where(
          and(
            eq(subscriptions.userId, userId),
            eq(subscriptions.name, sub.name)
          )
        )
        .limit(1);

      if (existingSub.length === 0) {
        await db.insert(subscriptions).values({
          userId,
          name: sub.name,
          merchant: sub.merchant,
          amount: String(sub.amount),
          currency: sub.currency || "NGN",
          frequency: sub.frequency,
          subcategory: sub.subcategory,
          isAutoDetected: true,
          firstSeenAt: sub.firstSeen,
          lastSeenAt: sub.lastSeen,
          nextExpectedAt: sub.nextExpected || undefined,
          confidence: String(sub.confidence),
          occurrenceCount: sub.occurrences.length,
        });
      }
    }

    // Track unknown account number transfers
    const accountTransfers = insertedTransactions.filter(
      (t) =>
        t.type === "debit" &&
        /\b\d{10}\b/.test(t.description) &&
        !t.merchant
    );

    for (const tx of accountTransfers) {
      const match = tx.description.match(/\b(\d{10})\b/);
      if (match) {
        const accountNumber = match[1];
        const txAmount = parseFloat(String(tx.amount));

        const existing = await db
          .select()
          .from(unknownAccounts)
          .where(
            and(
              eq(unknownAccounts.userId, userId),
              eq(unknownAccounts.accountNumber, accountNumber)
            )
          )
          .limit(1);

        if (existing.length === 0) {
          await db.insert(unknownAccounts).values({
            userId,
            accountNumber,
            totalSent: String(txAmount),
            transactionCount: 1,
          });
        } else {
          const e = existing[0];
          await db
            .update(unknownAccounts)
            .set({
              totalSent: String(parseFloat(String(e.totalSent)) + txAmount),
              transactionCount: (e.transactionCount || 0) + 1,
            })
            .where(eq(unknownAccounts.id, e.id));
        }
      }
    }

    // Balance validation
    let balanceValidated: boolean | undefined;
    let validationNotes: string | undefined;

    if (
      extractionResult.openingBalance !== undefined &&
      extractionResult.closingBalance !== undefined
    ) {
      const totalCredits = normalizedTransactions
        .filter((t) => t.type === "credit" && !t.isDuplicate)
        .reduce((s, t) => s + t.amount, 0);
      const totalDebits = normalizedTransactions
        .filter((t) => t.type === "debit" && !t.isDuplicate)
        .reduce((s, t) => s + t.amount, 0);

      const calculatedClose =
        extractionResult.openingBalance + totalCredits - totalDebits;
      const diff = Math.abs(
        calculatedClose - extractionResult.closingBalance
      );

      balanceValidated = diff < 1; // Allow 1 naira rounding difference
      validationNotes = balanceValidated
        ? "Statement balance validated successfully"
        : `Balance discrepancy of ₦${diff.toFixed(2)} detected`;
    }

    // Mark complete
    await db
      .update(statements)
      .set({
        status: "completed",
        extractedCount: insertedTransactions.length,
        balanceValidated,
        validationNotes,
        processingCompletedAt: new Date(),
      })
      .where(eq(statements.id, statementId));

    console.log(
      `Statement ${statementId} processed: ${insertedTransactions.length} transactions`
    );
  } catch (err) {
    console.error(`Statement ${statementId} processing failed:`, err);
    await db
      .update(statements)
      .set({
        status: "failed",
        processingError:
          err instanceof Error ? err.message : "Unknown error occurred",
      })
      .where(eq(statements.id, statementId));
  }
}
